tests/experimental/wgit/test_signal_sparsity.py

```diff
@@ -0,0 +1,142 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the BSD license found in the
+# LICENSE file in the root directory of this source tree.
+
+import random
+
+import pytest
+import torch
+
+from fairscale.experimental.wgit.signal_sparsity import SignalSparsity, _get_k_for_topk, _top_k_total_size
+
+TRIALS = 5
+TENSOR_DIM_LOWER = 40
+TENSOR_DIM_UPPER = 100
+DIM_LIST = [None, 0, 1]
+SPARSITY_ATOL = 8e-2
+FFT_RECONS_ATOL = 2e-3
+
+
+def test_validate_conf():
+    """Tests the config validation for the Signal Sparsity class."""
+
+    def kwargs(vals_list):
+        """Maps the values in input vals_list to the keys in arg_key_list"""
+
+        arg_key_list = [
+            "sst_top_k_element",
+            "sst_top_k_percent",
+            "sst_top_k_dim",
+            "dst_top_k_element",
+            "dst_top_k_percent",
+            "dst_top_k_dim",
+        ]
+        return dict(zip(arg_key_list, vals_list))
+
+    # Validate assertion error is raised when, either
+    # 1. both of sst (or dst) percent and element is set to values (not None).
+    # 2. both of sst (or dst) percent and element is set to None.
+    element = random.randint(0, TENSOR_DIM_UPPER)
+    percent = random.uniform(0, 100)
+    dim = random.randint(-1, 3)
+
+    pytest.raises(AssertionError, SignalSparsity, **kwargs([element, percent, dim, element, None, dim]))
+    pytest.raises(AssertionError, SignalSparsity, **kwargs([element, None, dim, element, percent, dim]))
+    pytest.raises(AssertionError, SignalSparsity, **kwargs([None, None, dim, element, None, dim]))
+    pytest.raises(AssertionError, SignalSparsity, **kwargs([element, None, dim, None, None, dim]))
+
+
+def test_sst_sparsity_level():
+    """Tests the dense_to_sst method of SignalSparsity in terms of generating SST at the correct
+    sparsity level.
+    """
+
+    def get_sst(sst_top_k_percent, dst_top_k_percent=100):
+        sparser_2d = SignalSparsity(sst_top_k_percent=sst_top_k_percent, dst_top_k_percent=100)
+        sst = sparser_2d.dense_to_sst(tensor)
+        return sst
+
+    for _ in range(TRIALS):
```

Comment:
ok, for the rest of test, I suggest we can simplify them and make them more targeted.

1. test for a fix input, expect a fixed output. Say a 5x5 input (not random, fixed in the test code, like 1,2,3,4,5, then 5,4,3,2,1, etc, etc. Then get the sst at element=2/percent=50, compare the output with a fixed expected output. This should be deterministic on all torch versions too. (we will find out)
2. test for some corner cases:
2.1. test a 2D input with dim=None, this should catch the bug where reshape not done on the result
2.2. test a 3D input with dim=1, which is the middle dim
2.3. more cases if we find more bugs in the future.

Again, in all case, fixed input and fixed expected output is good to test this kind of code. You can still do random test like you have done below, but it may not be efficient in catching bugs. 

In general, randomness in testing is not very good. If you have to have it, at least log all the random info and/or have a fixed seed so that random failure can be reproduced with a seed or with the logged info.

In general, testing should be tight, i.e. checking for exact expected value is better since any bug cause small change will be caught. Clearly, tight testing need determinism.

Finally, testing should cover all code paths. i.e. high code coverage. There are tools for measuring code coverage. But mentally you can check if your tests triggers all code paths to be executed when the code is small enough.

Reply:
Thank you so much for outlining the testing strategy. WIll go ahead with this. I think some of your sub-points within 2. are covered in line 133 of the tests, where we test dims for linear. Will go ahead and integrate the rest of the testing strategy you suggested!